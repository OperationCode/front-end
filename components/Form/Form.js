import { connect } from 'formik';

export const Form = connect(({ formik: { handleReset, handleSubmit }, ...props }) => (
  <form onReset={handleReset} onSubmit={handleSubmit} noValidate {...props} />
));

Form.displayName = 'Form';
