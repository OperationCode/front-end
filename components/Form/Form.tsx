import { connect } from 'formik';
import type { DetailedHTMLProps, FormHTMLAttributes, FunctionComponent } from 'react';

const FormikConnectedForm = connect(({ formik: { handleReset, handleSubmit }, ...props }) => (
  <form onReset={handleReset} onSubmit={handleSubmit} noValidate {...props} />
)) as FunctionComponent<DetailedHTMLProps<FormHTMLAttributes<HTMLFormElement>, HTMLFormElement>>;

FormikConnectedForm.displayName = 'FormikConnectedForm';

export default FormikConnectedForm;
