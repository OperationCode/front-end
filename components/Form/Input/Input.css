.field {
  margin: 1rem;
}

.Input {
  border: 1px solid rgba(var(--rgbValuesSecondary), 0.5);
  border-radius: 3px;
  font-size: 1.125rem;
  margin-top: 0.25rem;
  margin-bottom: 1rem;
  padding: 0.5rem;
  min-width: 200px;
}

.Input:focus {
  border-color: var(--primary);
  box-shadow: 0 0 1px 1px rgba(var(--rgbValuesPrimary), 0.75);
  outline: none;
}

.Input:disabled {
  opacity: 0.6;
}

.Input:disabled:hover {
  cursor: not-allowed;
}

.errorMessage {
  /* Alert's border is only 1px - not 2px (account for Input's left and right border) */
  max-width: calc(inherit + 2px);
}

.valid {
  border-color: var(--successDeep);
  box-shadow: 0 0 1px 1px var(--successDeep);
  outline: none;
}

.invalid {
  border-color: var(--errorDeep);
  box-shadow: 0 0 1px 1px var(--errorDeep);
  outline: none;
}
