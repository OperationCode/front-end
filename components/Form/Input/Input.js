import React from 'react';
import { shape, string, number, object, objectOf, oneOfType, bool, oneOf } from 'prop-types';
import classNames from 'classnames';
import { ErrorMessage } from 'formik';
import { INPUT, INPUT_ERROR, INPUT_FEEDBACK_GROUPING } from 'common/constants/testIDs';
import Alert from 'components/Alert/Alert';
import Label from 'components/Form/Label/Label';
import styles from './Input.module.css';

Input.propTypes = {
  field: shape({
    name: string.isRequired,
  }).isRequired,
  form: shape({
    // TODO: Resolve why multiselects can end up with touched: { key: array }
    // see ThemedReactSelect as well
    // touched: objectOf(bool).isRequired,
    touched: object.isRequired,
    errors: objectOf(string),
  }).isRequired,
  isLabelHidden: bool,
  id: oneOfType([string, number]),
  label: string.isRequired,
  hasValidationStyling: bool,
  type: oneOf([
    'button',
    'checkbox',
    'color',
    'date',
    'datetime-local',
    'email',
    'file',
    'hidden',
    'image',
    'month',
    'number',
    'password',
    'radio',
    'range',
    'reset',
    'search',
    'submit',
    'tel',
    'text',
    'time',
    'url',
    'week',
  ]),
};

Input.defaultProps = {
  hasValidationStyling: true,
  isLabelHidden: false,
  id: '',
  type: 'text',
};

function Input({
  field: { name, value, ...field },
  form: { touched, errors },
  hasValidationStyling,
  id,
  isLabelHidden,
  label,
  type,
  ...props // input simply has too many possible attributes... we'd be redocumenting the web
  // See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Attributes_common_to_all_input_types
}) {
  const hasErrors = Boolean(errors[name]);
  const isLabelAfterInput = type === 'radio' || type === 'checkbox';
  const isLabelBeforeInput = !isLabelAfterInput;

  return (
    <div className={styles.field} data-testid={INPUT}>
      {isLabelBeforeInput && (
        <Label for={name} isHidden={isLabelHidden}>
          {label}
        </Label>
      )}

      <div className={styles.inputFeedbackGrouping} data-testid={INPUT_FEEDBACK_GROUPING}>
        <input
          {...field}
          {...props}
          className={classNames(styles.Input, hasValidationStyling, {
            [styles.valid]: touched[name] && !hasErrors && hasValidationStyling,
            [styles.invalid]: touched[name] && hasErrors && hasValidationStyling,
          })}
          id={id || name}
          name={name}
          type={type}
          value={value || ''}
        />

        <ErrorMessage
          name={name}
          render={message => {
            return hasErrors ? (
              <Alert className={styles.errorMessage} data-testid={INPUT_ERROR} type="error">
                {message}
              </Alert>
            ) : null;
          }}
        />
      </div>

      {isLabelAfterInput && (
        <Label for={name} isHidden={isLabelHidden} className={styles.labelAfterInput}>
          {label}
        </Label>
      )}
    </div>
  );
}

export default Input;
