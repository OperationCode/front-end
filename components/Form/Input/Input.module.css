.field {
  margin: 1rem;
}

.inputFeedbackGrouping {
  display: flex;
  flex-direction: column;
  align-items: stretch;
}

.Input {
  border: 1px solid rgba(var(--rgbValuesSecondary), 0.5);
  border-radius: 3px;
  font-size: 1.125rem;
  padding: 0.5rem;
  min-width: 200px;
}

.Input[type="number"] {
  inline-size: auto;
}

.Input:focus {
  border-color: var(--primary);
  box-shadow: 0 0 1px 1px rgba(var(--rgbValuesPrimary), 0.75);
  outline: none;
}

.Input:disabled {
  opacity: 0.6;
}

.Input:disabled:hover {
  cursor: not-allowed;
}

.errorMessage {
  /* Alert's border is only 1px - not 2px (account for Input's left and right border) */
  max-width: calc(100% + 2px);
  margin-top: 0.5rem;
  flex: 1;
  text-align: center;
}

.valid {
  border-color: var(--successDeep);
  box-shadow: 0 0 1px 1px var(--successDeep);
  outline: none;
}

.invalid {
  border-color: var(--errorDeep);
  box-shadow: 0 0 1px 1px var(--errorDeep);
  outline: none;
}

@media screen and (--desktop-viewport) {
  .inputFeedbackGrouping {
    position: relative;
  }

  .errorMessage {
    margin-top: 0;
    margin-left: 1rem;
    position: absolute;
    top: -2px;
    left: 100%;
    min-width: 150px;
    max-width: 300px;
    width: auto;
    padding: 0 10px;
    min-height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
  }
}
