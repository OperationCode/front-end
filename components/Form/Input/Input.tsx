import type { FieldInputProps, FormikHelpers, FormikState } from 'formik';
import { ErrorMessage } from 'formik';
import { INPUT, INPUT_ERROR, INPUT_FEEDBACK_GROUPING } from 'common/constants/testIDs';
import type { InputHTMLAttributes } from 'react';
import { cx } from 'common/utils/cva';
import Label from 'components/Form/Label/Label';
import Alert from 'components/Alert/Alert';

/** @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Attributes_common_to_all_input_types */
interface InputProps extends Omit<InputHTMLAttributes<HTMLInputElement>, 'disabled' | 'form'> {
  field: FieldInputProps<string>;
  form: FormikState<Record<string, string>> & FormikHelpers<Record<string, string>>;
  hasValidationStyling?: boolean;
  isDisabled?: boolean;
  isLabelHidden?: boolean;
  label: string;
}

function Input({
  className,
  isDisabled = false,
  field: { name, value, ...field },
  form: { touched, errors },
  hasValidationStyling = true,
  id,
  isLabelHidden = false,
  label,
  type = 'text',
  ...props
}: InputProps) {
  const hasErrors = Boolean(errors[name]);
  const isLabelAfterInput = type === 'radio';
  const isLabelBeforeInput = !isLabelAfterInput;

  return (
    <div className={className} data-testid={INPUT}>
      {isLabelBeforeInput && (
        <Label htmlFor={id || name} isHidden={isLabelHidden}>
          {label}
        </Label>
      )}

      <div
        className="flex flex-col items-stretch lg:relative"
        data-testid={INPUT_FEEDBACK_GROUPING}
      >
        <input
          {...field}
          {...props}
          className={cx(
            'border border-secondary/50 rounded-sm text-lg p-2',
            'disabled:opacity-60 hover:disabled:cursor-not-allowed min-w-48',
            'focus-visible:border-primary/50 focus-visible:shadow-xs focus-visible:shadow-primary/75 focus-visible:outline-none',
            {
              'border-success-deep shadow-xs shadow-success-deep outline-none':
                touched[name] && !hasErrors && hasValidationStyling,
              'border-error-deep shadow-xs shadow-error-deep':
                touched[name] && hasErrors && hasValidationStyling,
            },
          )}
          disabled={isDisabled}
          id={id || name}
          name={name}
          type={type}
          value={value || ''}
        />

        <ErrorMessage name={name}>
          {(message: string) => {
            return hasErrors ? (
              <Alert
                className={cx(
                  'max-w-full -mx-0.5 mt-2 flex-1',
                  'lg:mt-0 lg:ml-4 lg:absolute lg:top-0 lg:left-full',
                  'lg:min-w-36 lg:max-w-72 lg:w-auto',
                  'lg:py-0 lg:px-2.5 lg:min-h-full lg:flex lg:items-center lg:justify-center',
                )}
                data-testid={INPUT_ERROR}
                type="error"
              >
                {message}
              </Alert>
            ) : null;
          }}
        </ErrorMessage>
      </div>

      {isLabelAfterInput && (
        <Label htmlFor={name} isHidden={isLabelHidden}>
          {label}
        </Label>
      )}
    </div>
  );
}

export default Input;
