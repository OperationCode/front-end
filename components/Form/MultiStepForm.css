.MultiStepForm {
  width: auto;
}

.containerFormErrorMessage {
  display: flex;
  justify-content: center;
}

.buttonGrouping {
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-around;
  align-items: center;
  margin-left: auto;
  margin-right: auto;
  margin-top: 2rem;

  /*
   * Specific value just so outer edges of each button visually aligns with
   * the outer edges of inputs in our often-existing two-column forms
   */
  width: 92.5%;
}

.buttonGrouping button + button {
  margin-left: 1rem;
}

@media screen and (--small-viewport) {
  .buttonGrouping button {
    min-width: 100px;
  }
}

.forceButtonRight {
  justify-content: flex-end;
}
