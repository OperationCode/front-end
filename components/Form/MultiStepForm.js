/* eslint-disable unicorn/prevent-abbreviations */
import React, { useState } from 'react';
import { arrayOf, func, object } from 'prop-types';
import noop from 'lodash/noop';
import { Formik } from 'formik';
import { validStep } from 'common/constants/custom-props';
import Button from 'components/Button/Button';
import Form from 'components/Form/Form';
import Alert from 'components/Alert/Alert';
import ProgressIndicator from 'components/ProgressIndicator/ProgressIndicator';
import styles from './MultiStepForm.css';

MultiStepForm.propTypes = {
  // initialValues must be object where entire form's shape is described
  initialValues: object.isRequired,

  getErrorMessage: func.isRequired,
  onEachStepSubmit: func,
  onFinalSubmit: func.isRequired, // to be considered onSuccess
  steps: arrayOf(validStep).isRequired,
};

MultiStepForm.defaultProps = {
  onEachStepSubmit: noop,
};

export function MultiStepForm({
  steps,
  initialValues,
  onEachStepSubmit,
  onFinalSubmit,
  getErrorMessage,
}) {
  const [state, setState] = useState({
    errorMessage: '',
    stepNumber: 0,
  });

  const { errorMessage, stepNumber } = state;

  const isLastStep = () => {
    return stepNumber === steps.length - 1;
  };

  // We assume this method cannot be called on the last step
  const showNextStep = ({ setFieldTouched }) => {
    // TODO: Only untouch if value is '' or []
    const nextStepFieldNames = Object.keys(steps[stepNumber + 1].initialValues);
    nextStepFieldNames.forEach(fieldName => setFieldTouched(fieldName, false));
    setState(prevState => ({ ...prevState, stepNumber: prevState.stepNumber + 1 }));
  };

  // We assume this method cannot be called on the first step
  const showPreviousStep = ({ setFieldTouched }) => {
    // TODO: Only untouch if value is '' or []
    const previousStepFieldNames = Object.keys(steps[stepNumber - 1].initialValues);
    previousStepFieldNames.forEach(fieldName => setFieldTouched(fieldName, false));

    setState(prevState => ({ ...prevState, stepNumber: prevState.stepNumber - 1 }));
  };

  const handleError = error => {
    setState(prevState => ({ ...prevState, errorMessage: getErrorMessage(error) }));
  };

  const handleSubmit = async (values, formikBag) => {
    if (errorMessage) {
      // reset error message each submit
      setState(prevState => ({ ...prevState, errorMessage: '' }));
    }

    try {
      await onEachStepSubmit(values);

      const currentStepSubmitHandler = steps[stepNumber].submitHandler;
      await currentStepSubmitHandler(values);

      if (isLastStep()) {
        await onFinalSubmit(values);
        formikBag.setSubmitting(false);
        formikBag.resetForm();
      } else {
        formikBag.setSubmitting(false);
        showNextStep(formikBag);
      }
    } catch (error) {
      formikBag.setSubmitting(false);
      handleError(error);
    }
  };

  const CurrentStep = steps[stepNumber];
  const isFirstStep = stepNumber === 0;

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={CurrentStep.validationSchema}
      onSubmit={handleSubmit}
      render={formikBag => (
        <Form className={styles.MultiStepForm} onSubmit={formikBag.handleSubmit}>
          <h3 className={styles.centerAligned}>{CurrentStep.title}</h3>

          <ProgressIndicator stepNumber={stepNumber} totalSteps={steps.length} />

          <CurrentStep {...formikBag} />

          <div className={styles.errorMessage}>
            {errorMessage && <Alert type="error">{errorMessage}</Alert>}
          </div>

          <div className={styles.buttonGrouping}>
            {!isFirstStep && (
              <Button
                theme="secondary"
                disabled={formikBag.isSubmitting}
                onClick={() => showPreviousStep(formikBag)}
                data-testid="Previous Step Button"
              >
                ← Previous
              </Button>
            )}

            {isLastStep() ? (
              <Button
                type="submit"
                theme="secondary"
                disabled={formikBag.isSubmitting}
                data-testid="Submit Multi-Step Form"
              >
                Submit ✓
              </Button>
            ) : (
              <Button
                type="submit"
                theme="secondary"
                disabled={formikBag.isSubmitting}
                fullWidth={isFirstStep}
                data-testid="Submit Step Button"
              >
                Next →
              </Button>
            )}
          </div>
        </Form>
      )}
    />
  );
}

export default MultiStepForm;
