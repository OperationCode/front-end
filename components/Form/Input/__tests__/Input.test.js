/* eslint-env jest */
import React from 'react';
import { Formik, Field } from 'formik';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import wait from 'test-utils/wait';
import { shallow, mount } from 'enzyme';

import Form from '../../Form';
import Input from '../Input';

describe('Input', () => {
  const requiredProps = {
    field: {
      name: 'someInputName',
    },
    form: { touched: { someInputName: false }, errors: { someInputName: '' } },
    label: 'Some Input:',
  };

  it('should render with required props', () => {
    createSnapshotTest(<Input {...requiredProps} />);
  });

  it('should render with label, even if hidden', () => {
    const wrapper = shallow(<Input {...requiredProps} isLabelHidden />);

    expect(wrapper).toContainExactlyOneMatchingElement('Label');
  });

  it('should display an error message when a required field is touched', async () => {
    const validate = () => ({ test: 'Required' });

    const wrapper = mount(
      <Formik validate={validate}>
        <Form>
          <Field id="test" name="test" label="label" component={Input} />,
        </Form>
      </Formik>,
    );

    wrapper.find('input').simulate('blur'); // trigger validation
    await wait();
    wrapper.update();
    expect(wrapper.find('Alert')).toHaveText('Required');
  });

  it('should render the label after input, but only when input type is radio or checkbox', () => {
    const checkboxInput = shallow(<Input {...requiredProps} type="checkbox" />);
    const radioInput = shallow(<Input {...requiredProps} type="radio" />);

    const checkboxField = checkboxInput.find('.field');
    const radioField = radioInput.find('.field');

    expect(checkboxField.childAt(0)).toContainExactlyOneMatchingElement('input');
    expect(checkboxField.childAt(1)).toContainExactlyOneMatchingElement('Label');

    expect(radioField.childAt(0)).toContainExactlyOneMatchingElement('input');
    expect(radioField.childAt(1)).toContainExactlyOneMatchingElement('Label');

    const otherInputTypes = [
      'button',
      'color',
      'date',
      'datetime-local',
      'email',
      'file',
      'hidden',
      'image',
      'month',
      'number',
      'password',
      'range',
      'reset',
      'search',
      'submit',
      'tel',
      'text',
      'time',
      'url',
      'week',
    ];

    expect(otherInputTypes).not.toContain('radio');
    expect(otherInputTypes).not.toContain('checkbox');

    otherInputTypes.forEach(inputType => {
      const input = shallow(<Input {...requiredProps} type={inputType} />);

      expect(input.find('.field').childAt(0)).toContainExactlyOneMatchingElement('Label');
      expect(input.find('.field').childAt(1)).toContainExactlyOneMatchingElement('input');
    });
  });

  // TODO: Add test for valid/invalid styles based on touched/errors
});
