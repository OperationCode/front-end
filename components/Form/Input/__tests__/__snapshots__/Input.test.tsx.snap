// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should render with required props 1`] = `
<div
  className="field"
  data-testid="INPUT"
>
  <label
    className="text-themeSecondary text-sm"
    data-testid="LABEL"
    htmlFor="someInputName"
  >
    Some Input:
  </label>
  <div
    className="inputFeedbackGrouping"
    data-testid="INPUT_FEEDBACK_GROUPING"
  >
    <input
      className="Input"
      id="someInputName"
      name="someInputName"
      onBlur={[MockFunction spy]}
      onChange={[MockFunction spy]}
      type="text"
      value=""
    />
  </div>
</div>
`;
